src/main/kotlin/com/terraformation/backend/file/FileService.kt

```diff
@@ -0,0 +1,127 @@
+package com.terraformation.backend.file
+
+import com.terraformation.backend.auth.currentUser
+import com.terraformation.backend.db.FileNotFoundException
+import com.terraformation.backend.db.default_schema.FileId
+import com.terraformation.backend.db.default_schema.tables.daos.FilesDao
+import com.terraformation.backend.db.default_schema.tables.pojos.FilesRow
+import com.terraformation.backend.db.default_schema.tables.references.FILES
+import com.terraformation.backend.file.model.FileMetadata
+import com.terraformation.backend.log.perClassLogger
+import java.io.IOException
+import java.io.InputStream
+import java.net.URI
+import java.nio.file.FileAlreadyExistsException
+import java.nio.file.NoSuchFileException
+import java.time.Clock
+import javax.inject.Named
+import org.jooq.DSLContext
+
+/**
+ * Manages storage of files including metadata. In this implementation, files are stored on the
+ * filesystem and metadata in the database.
+ */
+@Named
+class FileService(
+    private val dslContext: DSLContext,
+    private val clock: Clock,
+    private val filesDao: FilesDao,
+    private val fileStore: FileStore,
+    private val thumbnailStore: ThumbnailStore,
+) {
+  private val log = perClassLogger()
+
+  @Throws(IOException::class)
+  fun storeFile(
+      category: String,
+      data: InputStream,
+      size: Long,
+      metadata: FileMetadata,
+      insertChildRows: (FileId) -> Unit
+  ): FileId {
+    val storageUrl = fileStore.newUrl(clock.instant(), category, metadata.contentType)
+
+    try {
+      fileStore.write(storageUrl, data)
+
+      val filesRow =
+          FilesRow(
+              contentType = metadata.contentType,
+              createdTime = clock.instant(),
+              createdBy = currentUser().userId,
+              fileName = metadata.filename,
+              modifiedBy = currentUser().userId,
+              modifiedTime = clock.instant(),
+              size = size,
+              storageUrl = storageUrl,
+          )
+
+      dslContext.transaction { _ ->
+        filesDao.insert(filesRow)
+        insertChildRows(filesRow.id!!)
+      }
+
+      return filesRow.id!!
+    } catch (e: FileAlreadyExistsException) {
+      // Don't delete the existing file
+      throw e
+    } catch (e: Exception) {
+      try {
+        fileStore.delete(storageUrl)
+      } catch (ignore: NoSuchFileException) {
+        // Swallow this; file is already deleted
+      }
+      throw e
+    }
+  }
+
+  @Throws(IOException::class)
+  fun readFile(
+      fileId: FileId,
+      maxWidth: Int? = null,
+      maxHeight: Int? = null,
+  ): SizedInputStream {
+    return if (maxWidth != null || maxHeight != null) {
+      thumbnailStore.getThumbnailData(fileId, maxWidth, maxHeight)
```

Comment:
just a thought, we could remove width/height semantics and directly call thumbnail store where needed.. or have a derived photo service (but that's certainly overkill)

Reply:
Having separate thumbnail endpoints is a fair suggestion, but I think it's better if this PR is a pure refactor that doesn't also introduce client-visible changes.

Reply:
Makes sense, totally agree 👍